activerecord/lib/active_record/railtie.rb

```diff
@@ -98,6 +98,22 @@ class Railtie < Rails::Railtie # :nodoc:
       end
     end
 
+    initializer "active_record.asynchronous_queries_executor" do
+      if concurrency = config.active_record.delete(:asynchronous_queries_concurrency)
+        unless config.active_record.asynchronous_queries_executor
+          config.active_record.asynchronous_queries_executor = if concurrency == 0
+            Concurrent::ImmediateExecutor.new
+          else
+            Concurrent::ThreadPoolExecutor.new(
+             min_threads: 0,
+             max_threads: concurrency,
+             max_queue: concurrency * 4,
+             fallback_policy: :caller_runs
```

Comment:
I think I misunderstood the direction you mentioned going in for this when we spoke earlier today.

By putting this in the initializer and having a global setting for `max_threads` we can't configure the thread pool per database whereas before it was using the pool size from the database config for max threads.

I was going to send a PR that allows the db config to set these values so each db can control it's own parallelism because that's how we plan to use it at GitHub. This implementation would make it impossible to control the thread pool execution per database.

Reply:
I see. The ThreadPool ownership is really something I still haven't managed to find a clear cut solution for this.

In GItHub case IIRC you are sharing vertically (ModelA is on one shard, ModelB on another), which means for you it's interesting to have a concurrency limit per database.

In Shopify case we're sharding horizontally (ModelA and ModelB are on X instances of the same schema), which means for us we'd end up with dozens if not hundreds of threads sitting idle most of the time.

Maybe we need an actual API to select the thread pool? Or maybe we need a more complex executor that would allow to have a single ThreadPool but with multiple task queues with each their own concurrency limit?

I'll have to sleep on this.

Reply:
An important use case (that we probably both share?) would be a different concurrency between reader (replica) and writer databases.

> maybe we need a more complex executor that would allow to have a single ThreadPool but with multiple task queues with each their own concurrency limit

This feels ideal but probably tricky to get right without starvation.

Reply:
Hum, right. Maybe we should stick with the current solution, but just make it easy to replace the strategy?